src/components/CPU/BreakpointPanel.js

```diff
@@ -0,0 +1,167 @@
+import React, {useCallback, useEffect, useState} from 'react';
+import PropTypes from 'prop-types';
+import '../ext/react-modal.css';
+import './BreakpointPanel.css';
+import {toString08X} from "../../utils/format";
+import {ContextMenu, ContextMenuTrigger, MenuItem} from "react-contextmenu";
+import listeners from "../../utils/listeners";
+
+export default function BreakpointPanel(props) {
+	const [cpuBreakpoints, setCpuBreakpoints] = useState([]);
+	const [memoryBreakpoints, setMemoryBreakpoints] = useState([]);
+
+	const fetchBreakpoints = useCallback(() => {
+		props.ppsspp.send({
+			event: 'cpu.breakpoint.list',
+		}).then(({breakpoints}) => {
+			setCpuBreakpoints(breakpoints);
+		});
+		props.ppsspp.send({
+			event: 'memory.breakpoint.list',
+		}).then(({breakpoints}) => {
+			setMemoryBreakpoints(breakpoints);
+		});
+	}, [props.ppsspp]);
+
+	useEffect(() => {
+		fetchBreakpoints();
+		// TODO cpu.breakpoint.* not sent by PPSSPP?
+		const listeners_ = listeners.listen({
+			'cpu.stepping': () => {
+				fetchBreakpoints();
+			},
+			// TODO memory.breakpoint add/update/remove
+			'cpu.breakpoint.add': () => {
+				fetchBreakpoints();
+			},
+			'cpu.breakpoint.update': () => {
+				fetchBreakpoints();
+			},
+			'cpu.breakpoint.remove': () => {
+				fetchBreakpoints();
+			},
+		});
+		return () => listeners.forget(listeners_);
+	}, [fetchBreakpoints]);
+
+	return (
+		<React.Fragment>
+			<ContextMenuTrigger id="addBreakpoint" holdToDisplay={-1}>
+				<table className="BreakpointPanel__table">
+					<thead>
+						{/* TODO Left align header, first column sizing, row selection */}
+						<tr>
+							<th>&nbsp;</th>
+							<th>Type</th>
+							<th>Offset</th>
+							<th>Size/Label</th>
+							<th>Opcode</th>
+							<th>Condition</th>
+							<th>Hits</th>
+						</tr>
+					</thead>
+					<tbody>
+						{renderMemoryBreakpoints(memoryBreakpoints)}
+						{renderCpuBreakpoints(cpuBreakpoints)}
+					</tbody>
+				</table>
+			</ContextMenuTrigger>
+			{/*<BreakpointModal*/}
+			{/*	ppsspp={props.ppsspp}*/}
+			{/*	currentThread={props.currentThread}*/}
+			{/*	isOpen={false}*/}
+			{/*	// onClose={this.closeEditBreakpoint}*/}
+			{/*	// breakpoint={this.state.editingBreakpoint}*/}
+			{/*	// initialOverrides={this.state.creatingBreakpoint}*/}
+			{/*/>*/}
+			{renderAddBreakpointContextMenu()}
+			{renderEditBreakpointContextMenu()}
+		</React.Fragment>
+	);
+}
+
+function renderAddBreakpointContextMenu() {
+	return (
+		<ContextMenu id="addBreakpoint">
+			<MenuItem>
+				Add New
+			</MenuItem>
+		</ContextMenu>
+	);
+}
+
+function renderEditBreakpointContextMenu() {
+	// TODO probably can't do checked MenuItem so maybe render dynamic text depending on breakpoint state: Enable/Disable

```

Comment:
See src/components/CPU/DisasmContextMenu.js and src/components/CPU/DisasmLine.js for an example of a ContextMenu that varies based on the selected item.  Specifically `<MenuItem disabled={!followBranch}` and `mapData`.  We could even have it as one context menu that varies based on trigger.

That said, maybe we change to Toggle Break.  I've always considered Enable/Disable a bit unclear considering you can have a log-only breakpoint and "disabling" it doesn't stop it from logging.

-[Unknown]

Reply:
> That said, maybe we change to Toggle Break. I've always considered Enable/Disable a bit unclear considering you can have a log-only breakpoint and "disabling" it doesn't stop it from logging.

Interesting, I always thought that the checkbox disables logging too (but I rarely used logging breakpoints). In that case I agree on calling it Toggle Break.

True, it can be simplified as a single menu.

I appreciate the other review comments too and I kind of didn't expect you would be going through this more thoroughly. Just wanted to get feedback on the two things from the first post. Anyway, I'll continue working on this.

Reply:
Implemented as single menu.